src/net_processing.cpp

```diff
@@ -186,6 +186,38 @@ struct QueuedBlock {
     std::unique_ptr<PartiallyDownloadedBlock> partialBlock;
 };
 
+template <size_t N>
+class TimeOffsets
+{
+private:
+    static_assert(N > 0, "TimeOffsets has to sample at least one element");
+
+    mutable Mutex m_mutex;
+    std::array<int64_t, N> m_offsets GUARDED_BY(m_mutex){};
+    size_t m_index GUARDED_BY(m_mutex){0};
+
+public:
+    void Add(int64_t offset) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        AssertLockNotHeld(m_mutex);
+        LOCK(m_mutex);
+        m_offsets[m_index] = offset;
+        m_index = (1 + m_index) % N;
+    }
+    int64_t Median() const EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        AssertLockNotHeld(m_mutex);
+        std::array<int64_t, N> sorted_copy = WITH_LOCK(m_mutex, return m_offsets);
+        std::sort(sorted_copy.begin(), sorted_copy.end());
+
+        if constexpr (N % 2 == 0) {
+            return (sorted_copy[N / 2 - 1] + sorted_copy[N / 2]) / 2;
```

Comment:
The idea is that `m_index` is always set to the next index in `m_offsets` that gets written to (so a value out of `[0, N)`) and we just assume that the initial N samples were 0, which is why the median is immediately computed out of N.

I could change it to always increase `m_index` instead and add new offsets by doing `m_offsets[m_index % N] = new_offset`, in which case your suggestion would allow us to compute the median more "accurately" while we haven't seen N samples yet. I'm not sure though if that is an improvement,. e.g. do we want to issue a warning if only the first outbound connection has a offset larger than our threshold?

Reply:
>do we want to issue a warning if only the first outbound connection has a offset larger than our threshold?

This is the main question I guess.
So, with the current code, having 4 peers would never trigger it, right? Because median will always be 0.